project-chip/connectedhomeip
# Add ability to send required events for Door Lock app

#### Problem
* Fixes #19807 by adding ability to send more events for Linux Door Lock app.

#### Change overview
1. Added ability to send commands to the lock-app over the named pipe. See [README.md](https://github.com/project-chip/connectedhomeip/blob/68769c9e34fd84f70c7893ba7e251a9f7fd6e75f/examples/lock-app/linux/README.md) for more details.
1. Now Lock App for Linux is able to restrict access for users of Schedule Restricted User based on schedules.

#### Testing
* Send `SetDoorState` command over the named pipe (see [README.md](https://github.com/project-chip/connectedhomeip/blob/68769c9e34fd84f70c7893ba7e251a9f7fd6e75f/examples/lock-app/linux/README.md)) and verify that the door state change event is sent by Lock App.
* Send `SendDoorLockAlarm` command over the named pipe (see [README.md](https://github.com/project-chip/connectedhomeip/blob/68769c9e34fd84f70c7893ba7e251a9f7fd6e75f/examples/lock-app/linux/README.md)) and verify that the Door Lock Alarm event with `LockJammed` type is sent by Lock App.
* Verify schedule support:
  1. Create a credential with user of Schedule Restricted Type:
     ```
     ? chip-tool doorlock set-credential 0 '{ "credentialType" : 1 , "credentialIndex" : 1 }' "123456" null null 8 1 1 --timedInteractionTimeoutMs 1000
     ```
  2. Verify that it is impossible to lock/unlock the door 
     ```
     ? chip-tool doorlock unlock-door 1 1 --timedInteractionTimeoutMs 1000 --PinCode "123456"
     ...
     [1658856707540] [64241:10356960] CHIP: [TOO] Run command failure: IM Error 0x00000501: General error: 0x01 (FAILURE)
     ```
  3. Verify the `LockOperationError` event is sent by Lock app:
     ```
     ? chip-tool doorlock read-event lock-operation-error 1 1
     ...
     [1658856750606] [64404:10358982] CHIP: [TOO] Endpoint: 1 Cluster: 0x0000_0101 Event 0x0000_0003
     [1658856750606] [64404:10358982] CHIP: [TOO]     Event number: 1900550
     [1658856750606] [64404:10358982] CHIP: [TOO]     Priority: Critical
     [1658856750606] [64404:10358982] CHIP: [TOO]     Timestamp: 2581137538
     [1658856750606] [64404:10358982] CHIP: [TOO]   LockOperationError: {
     [1658856750606] [64404:10358982] CHIP: [TOO]     LockOperationType: 1
     [1658856750606] [64404:10358982] CHIP: [TOO]     OperationSource: 7
     [1658856750606] [64404:10358982] CHIP: [TOO]     OperationError: 3
     [1658856750606] [64404:10358982] CHIP: [TOO]     UserIndex: 1
     [1658856750606] [64404:10358982] CHIP: [TOO]     FabricIndex: 1
     [1658856750606] [64404:10358982] CHIP: [TOO]     SourceNode: 112233
     [1658856750606] [64404:10358982] CHIP: [TOO]     Credentials: 1 entries
     [1658856750606] [64404:10358982] CHIP: [TOO]       [1]: {
     [1658856750606] [64404:10358982] CHIP: [TOO]         CredentialType: 1
     [1658856750606] [64404:10358982] CHIP: [TOO]         CredentialIndex: 1
     [1658856750606] [64404:10358982] CHIP: [TOO]        }
     [1658856750606] [64404:10358982] CHIP: [TOO]    }
     ...
     ```
  4. Create the Week Day schedule that is currently in action: 
     ```
     ? chip-tool doorlock set-week-day-schedule 1 1 0x04 10 00 21 00 1 1 --timedInteractionTimeoutMs 1000
     ```
  5. Verify that the lock can be unlocked now: 
     ```
     ? chip-tool doorlock unlock-door 1 1 --timedInteractionTimeoutMs 1000 --PinCode "123456"
     ```
  6. Do the same operations for Year Day Schedules.
* Re-ran all the existing lock YAML tests.
